src/node/rpc/member_frontend.h

```diff
@@ -1332,16 +1332,61 @@ namespace ccf
         }
 
         const auto in = params.get<Propose::In>();
-        const auto proposal_id =
-          fmt::format("{:02x}", fmt::join(caller_identity.request_digest, ""));
 
-        Proposal proposal(in.script, in.parameter, caller_identity.member_id);
+        if (!consensus)
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            "No consensus available.");
+        }
+
+        std::string proposal_id;
 
+        if (consensus->type() == ConsensusType::CFT)
+        {
+          auto root_at_read = ctx.tx.get_root_at_read_version();
+          if (!root_at_read.has_value())
+          {
+            return make_error(
+              HTTP_STATUS_INTERNAL_SERVER_ERROR,
+              ccf::errors::InternalError,
+              "Proposal failed to bind to state.");
+          }
+
+          // caller_identity.request_digest is set when getting the
+          // MemberSignatureAuthnIdentity identity. The proposal id is a digest
+          // of the root of the state tree at the read version and the request
+          // digest.
+          std::vector<uint8_t> acc(
+            root_at_read.value().h.begin(), root_at_read.value().h.end());
+          acc.insert(
+            acc.end(),
+            caller_identity.request_digest.begin(),
+            caller_identity.request_digest.end());
+          const crypto::Sha256Hash proposal_digest(acc);
+          proposal_id = proposal_digest.hex_str();
+        }
+        else
+        {
+          proposal_id = fmt::format(
+            "{:02x}", fmt::join(caller_identity.request_digest, ""));
+        }
+
+        Proposal proposal(in.script, in.parameter, caller_identity.member_id);
         auto proposals = ctx.tx.rw(this->network.proposals);
-        proposals->put(proposal_id, proposal);
+        // The existence of this check guarantees that we always enter this
+        // block because it creates a read-dependency on that proposal_id, and
+        // so any other request with an identical digest and the same
+        // read_version will conflict, be re-based, acquire a new read_version
+        // and therefore a new proposal_id.
+        if (!proposals->has(proposal_id))
```

Comment:
This behaviour is surprisingly subtle. Even reading the comment and dicsussion, I spent a while down the "what if this check fails" path (and so we never write and so this whole proposal becomes potentially read-only and so we get a duplicate proposal ID?) before convincing myself that can never happen.

What we're really saying is "the other case is a hash collision", so I think we should invert this test and document that:
```
// Introduce a read dependency, so that if identical proposal creations are in-flight
// and reading at the same version, the later conflicts and is re-executed. If we ever
// produce a proposal ID which already exists, we have a hash collision
if (proposals->hash(proposal_id))
{
 return make_error("HASH COLLISION: The sky is falling");
}

proposals->put(...)
record...
```
(proposed comment wording is inspirational and unmaintained)

Reply:
What I had in mind was that dependency handling could be broken in the KV. The defensive and correct thing to do is to avoid overwriting the key, since the proposal is structurally identical. Adding an assert() so this gets caught in debug would be better of course.

You're right that a hash collision could happen here (and in many other places). I do not know how to rate the relative probabilities of that, versus dependency handling breaking, but I'm happy to make the change you suggest if it improves readability.

Technically, we could read and compare the actual proposal contents, and work out which it is. I am not convinced it is useful.